rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,94 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
```

Comment:
We would still need a way to ensure this isn't applied arbitrarily in nixpkgs. We need CI and potentially nix features to prohibit usage of this feature and eventually newer nix versions until they can be assumed "adopted". This is the same discussion that we are having regarding bumping the minimal nix version requirement in nixpkgs.

Reply:
I think this is the exact same underlying issue, as @tazjin suggests, making the language version explicit would allow addressing it independent of C++ Nix. I think this is covered by these arguments:
- [makes language version explicit](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R43)
- [changing the language should be encoded in development culture](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R49-R50)
- [encoding the language version provides compatibility information to third-party evaluators](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R65-R67)